src/crypto/tls/ech_config.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2020 Cloudflare, Inc. All rights reserved. Use of this source code
+// is governed by a BSD-style license that can be found in the LICENSE file.
+
+package tls
+
+import (
+	"crypto/tls/internal/hpke"
+	"fmt"
+	"io"
+
+	"golang.org/x/crypto/cryptobyte"
+)
+
+// ECHConfig represents an ECH configuration.
+type ECHConfig struct {
+	pk  hpke.KEMPublicKey
+	raw []byte
+
+	// Parsed from raw
+	version           uint16
+	rawPublicName     []byte
+	rawPublicKey      []byte
+	kemId             uint16
+	suites            []echCipherSuite
+	maxNameLen        uint16
+	ignoredExtensions []byte
+}
+
+// UnmarshalECHConfigs parses a sequence of ECH configurations.
+func UnmarshalECHConfigs(raw []byte) ([]ECHConfig, error) {
+	var err error
+	var config ECHConfig
+	configs := make([]ECHConfig, 0)
+
+	s := cryptobyte.String(raw)
+	var t cryptobyte.String
+	if !s.ReadUint16LengthPrefixed(&t) || !s.Empty() {
+		return nil, fmt.Errorf("error parsing configs")
+	}
+	raw = raw[2:]
+
+	for !t.Empty() {
+		l := len(t)
+		if !echReadConfig(&t, &config) {
+			return nil, fmt.Errorf("error parsing config")
+		}
+		n := l - len(t)
+		config.pk, err = echUnmarshalHpkePublicKey(config.rawPublicKey, config.kemId)
+		if err != nil {
+			return nil, err
+		}
+		config.raw = raw[:n]
+		raw = raw[n:]
+		configs = append(configs, config)
+	}
+	return configs, nil
+}
+
+func echUnmarshalConfig(raw []byte) (*ECHConfig, error) {
+	var err error
+	s := cryptobyte.String(raw)
+	config := new(ECHConfig)
+	if !echReadConfig(&s, config) || !s.Empty() {
+		return nil, fmt.Errorf("error parsing config")
+	}
+	config.pk, err = echUnmarshalHpkePublicKey(config.rawPublicKey, config.kemId)
+	if err != nil {
+		return nil, err
+	}
+	config.raw = raw
+	return config, nil
+}
+
+func echReadConfig(s *cryptobyte.String, config *ECHConfig) bool {
+	if !s.ReadUint16(&config.version) { // version
+		return false
+	}
+
+	// Ensure we know how to proceed before attempting to parse the
+	// configuration contents.
+	if config.version != extensionECH {
+		return false
+	}
+
+	var contents cryptobyte.String // length + contents
+	if !s.ReadUint16LengthPrefixed(&contents) {
+		return false
+	}
+
+	var t cryptobyte.String
+	if !contents.ReadUint16LengthPrefixed(&t) || // contents.public_name
```

Comment:
How so? Operands to `||` are evaluated left-to-right, there cannot be reordering since it has side-effects.
This pattern to read a buffer and fail on the first failure is also common for users of `cryptobyte`.

Specifically for Go: https://golang.org/ref/spec#Logical_operators

Reply:
where in the spec says that the left-to-right order is enforced or guaranteed?

Reply:
Spec says it here: https://golang.org/ref/spec#Operator_precedence

Reply:
that's right